### PR TITLE
Create Posting CRUD Methods and GraphQL Resolvers

### LINKED ISSUE #50: Create CRUD services and GraphQL resolvers for Postings

### PR DESCRIPTION
**Pull request by albertlai431**
## Ticket link

Closes #50 

## Implementation description
- Implement CRUD methods:
  - getPosting
  - getPostings
  - createPosting
  - updatePosting
  - deletePosting
- Create corresponding GraphQL resolvers and DTOs
- Modified Prerequisites Model in schema as requested in ticket and generated associated prisma migration file 

## Steps to test
1. Start docker and run `docker-compose up --build`
2. We need row(s) in the skills, branches, and employees tables for testing and will need to create those with raw SQL for now. Run the following commands:
```bash
docker exec -it sistering_db /bin/bash
psql -U postgres -d sistering
INSERT INTO branches (name) VALUES ('Branch1');
INSERT INTO skills (name) VALUES ('Skill1');
INSERT INTO employees (userId, branch_id) VALUES ('1','1');
```
3. Run `docker exec -it sistering_backend /bin/bash -c "npx prisma generate"` and run the lastest migration files using `npx prisma migrate deploy`
4. Go to http://localhost:5000/graphql and verify that all GraphQL queries and mutations introduced in this ticket work as expected

## What should reviewers focus on?
* Verify that GraphQL endpoints work as expected

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/services/implementations/postingService.ts [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+          },
+          skills: {
+            connect: posting.skills.map((skillId: string) => {
+              return { id: Number(skillId) };
+            }),
+          },
+          employees: {
+            connect: posting.employees.map((employeeId: string) => {
+              return { id: Number(employeeId) };
+            }),
```
Comment (sherryhli) [inline backend/services/implementations/postingService.ts:237]: Please see this [comment](https://github.com/uwblueprint/sistering/pull/63/files#r764513528) in @jiarucao and @daniel-sm-yu's PR, will need to clear existing records in the join tables before connecting with the new ids

File: backend/services/implementations/postingService.ts [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+      startDate: posting.startDate,
+      endDate: posting.endDate,
+      autoClosingDate: posting.autoClosingDate,
+      numVolunteers: posting.numVolunteers,
+    };
+  }
+
+  async deletePosting(postingId: string): Promise<string> {
+    try {
+      const deleteResult: Posting | null = await prisma.posting.delete({
```
Comment (sherryhli) [inline backend/services/implementations/postingService.ts:276]: The delete needs to be cascaded to all associated shifts
https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions

File: backend/prisma/schema.prisma [hunk 3]
```diff
@@ -20,11 +20,11 @@ model Entity {
...
-  id             Int            @id @default(autoincrement())
-  firstName      String         @map("first_name")
-  lastName       String         @map("last_name")
-  authId         String         @unique @map("auth_id")
-  role           Role
+  id        Int    @id @default(autoincrement())
+  firstName String @map("first_name")
+  lastName  String @map("last_name")
+  authId    String @unique @map("auth_id")
+  role      Role
```
Comment (sherryhli) [inline backend/prisma/schema.prisma:27]: Nit: can we revert all pure formatting changes in this file? It'll prevent some annoying merge conflicts for the others who are also modifying things here

File: backend/prisma/schema.prisma [hunk 4]
```diff
@@ -153,44 +149,45 @@ model Volunteer {
...
 
   @@map("employees")
 }
 
-model PostingOnSkill {
-  posting   Posting @relation(fields: [postingId], references: [id])
-  postingId Int     @map("postings_id")
-  skill     Skill   @relation(fields: [skillId], references: [id])
-  skillId   Int     @map("skills_id")
+// model PostingOnSkill {
```
Comment (sherryhli) [inline backend/prisma/schema.prisma:161]: Let's delete everything that's commented out

**Review by sherryhli — CHANGES_REQUESTED** [review]
Looking good so far! Will do a second review once this set of comments is addressed

File: backend/services/implementations/postingService.ts [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+    return {
+      id: String(shift.id),
+      postingId: String(shift.postingId),
+      startTime: shift.startTime,
+      endTime: shift.endTime,
+    };
+  });
+}
+
+function validateDate(date: string) {
```
[the remark this answers is not in the document]
Comment (briantu) [reply backend/services/implementations/postingService.ts]: Got it, I'll keep it like this for now then.

File: backend/services/implementations/postingService.ts [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+          include: {
+            branch: true,
+            shifts: true,
+            skills: true,
+            employees: true,
+          },
+        },
+      );
+      return await Promise.all(
+        postings.map(async (posting: PostingWithRelations) => {
```
[the remark this answers is not in the document]
Comment (briantu) [reply backend/services/implementations/postingService.ts]: You're right, fixed!

File: backend/prisma/schema.prisma (hunk 3 above)
Comment (briantu) [reply backend/prisma/schema.prisma:27]: I think the current formatting changes follow Prisma's format. There is a `prisma format` command which I ran, and nothing was changed in this file. I think maybe we keep it like this and stay consistent with this Prisma's style of formatting?

File: backend/prisma/schema.prisma [hunk 7]
```diff
@@ -48,36 +48,32 @@ enum PostingType {
...
+  numVolunteers   Int         @default(1) @map("num_volunteers")
 
   @@map("postings")
 }
 
 model Skill {
   id         Int                @id @default(autoincrement())
-  name       String             @unique
-  postings   PostingOnSkill[]
+  name       String
```
Comment (sherryhli) [inline backend/prisma/schema.prisma:75]: Why was the unique constraint removed?

File: backend/prisma/schema.prisma (hunk 3 above)
Comment (sherryhli) [reply backend/prisma/schema.prisma:27]: Yep, definitely want to keep to this format eventually! Just thought it'd make things easier for Jiaru and Daniel in terms of merge conflicts in the meantime (they are also modifying this file in their PR).

But yeah, on second thought it shouldn't be too bad to deal with. Let's keep it 👍

**Review by sherryhli — CHANGES_REQUESTED** [review]
Almost there, just a couple of minor things left!

**Review by sherryhli — APPROVED** [review]
LGTM! 🚢
(please coordinate the order of merging with the rest of the team as described in [Slack](https://uwblueprint.slack.com/archives/C02CN0ZAZDG/p1639538084009000))

**Comment by sherryhli** [comment]
Extra nit: can we rename these files to keep consistency with the other interface files?

`backend/services/interfaces/IPostingService.ts` -> `backend/services/interfaces/postingService.ts`
`backend/services/interfaces/ISkillService.ts` -> `backend/services/interfaces/skillService.ts`
`backend/services/interfaces/IEntityService.ts` -> `backend/services/interfaces/entityService.ts`

**Review by briantu — APPROVED** [review]

**Pull request merged**